File: python/kinesis_publisher.py
# Kinesis Aggregation/Deaggregation Libraries for Python
# 
# Copyright 2014, Amazon.com, Inc. or its affiliates. All Rights Reserved. 
# 
# Licensed under the Amazon Software License (the "License").
# You may not use this file except in compliance with the License.
# A copy of the License is located at
# 
#  http://aws.amazon.com/asl/
# 
# or in the "license" file accompanying this file. This file is distributed
# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
# express or implied. See the License for the specific language governing
# permissions and limitations under the License.

import six
import sys

# Verify that we have access to boto
try:
    import boto3
    import botocore.config
except ImportError:
    six.print_("The 'boto3' module is required to run this script. Use 'pip install boto3' to get it.", file=sys.stderr)
    sys.exit(1)

import random
import uuid

import aws_kinesis_agg.aggregator
    
# Used for generating random record bodies
ALPHABET = 'abcdefghijklmnopqrstuvwxyz'

kinesis_client = None
stream_name = None


def get_random_record(seq_num=0, desired_len=50):
    """Generate a random record to send to Kinesis.
    
    Args:
        seq_num - The sequence number to include in the data body. (int)
        desired_len - The total size (in bytes) of the desired record body. (int)
    Returns:
        A semi-random string of the form "RECORD <seq_num> <random_alphabet_chars>"
        to use to populate the body of a Kinesis record. (str)"""
    
    global ALPHABET
    
    partition_key = str(uuid.uuid4())
    explicit_hash_key = str(uuid.uuid4().int)
    
    raw_data = 'RECORD %d ' % seq_num
    while len(raw_data) < (desired_len-1):
        raw_data += ALPHABET[random.randrange(0, len(ALPHABET))]
        raw_data += '\n'
    
    return partition_key, explicit_hash_key, raw_data


def init_kinesis_client(region):
    """Create a boto3 Kinesis client for the given reason.
    
    Args:
        region_name - The name of the AWS region the Kinesis client will be configured for (e.g. us-east-1) (str)
    Returns:
        A boto3 Kinesis client object configured for the input region."""
    
    global kinesis_client
    
    config = botocore.config.Config()
    config.region_name = region
    config.connection_timeout = 60
    config.read_timeout = 60
    
    kinesis_client = boto3.client('kinesis', config=config)


def send_record(agg_record):
    """Send the input aggregated record to Kinesis via the PutRecord API.
    
    Args:
        agg_record - The aggregated record to send to Kinesis. (AggRecord)"""
    
    global kinesis_client, stream_name
    
    if agg_record is None:
        return
    
    partition_key, explicit_hash_key, raw_data = agg_record.get_contents()
    
    six.print_('Submitting record with EHK=%s NumRecords=%d NumBytes=%d' %
               (explicit_hash_key, agg_record.get_num_user_records(), agg_record.get_size_bytes()))
    try:
        kinesis_client.put_record(StreamName=stream_name,
                                  Data=raw_data,
                                  PartitionKey=partition_key,
                                  ExplicitHashKey=explicit_hash_key)
    except Exception as e:
        six.print_('Transmission Failed: %s' % e, file=sys.stderr)
    else:
        six.print_('Completed record with EHK=%s' % ehk)
    
    
if __name__ == '__main__':
        
    # For details on how to supply AWS credentials to boto3, see:
    # https://boto3.readthedocs.org/en/latest/guide/configuration.html
    
    RECORD_SIZE_BYTES = 1024
    RECORDS_TO_TRANSMIT = 1024
    
    if len(sys.argv) != 3:
        six.print_("USAGE: python kinesis_publisher.py <stream name> <region>", file=sys.stderr)
        sys.exit(1)
        
    stream_name = sys.argv[1]
    region_name = sys.argv[2]
    
    init_kinesis_client(region_name)
    kinesis_agg = aws_kinesis_agg.aggregator.RecordAggregator()
    kinesis_agg.on_record_complete(send_record)
    
    six.print_('Creating %d records...' % RECORDS_TO_TRANSMIT)
    for i in range(1, RECORDS_TO_TRANSMIT+1):
        
        pk, ehk, data = get_random_record(i, RECORD_SIZE_BYTES)
        kinesis_agg.add_user_record(pk, data, ehk)
    
    # Do one final flush & send to get any remaining records that haven't triggered a callback yet
    send_record(kinesis_agg.clear_and_get())

File: node/kinesisPublisher.js
/*
		Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.

    Licensed under the Amazon Software License (the "License"). You may not use this file except in compliance with the License. A copy of the License is located at

        http://aws.amazon.com/asl/

    or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and limitations under the License. 
 */

var AWS = require('aws-sdk');
var libPath = "./node_modules/aws-kpl-agg"
var RecordAggregator = require(libPath + '/RecordAggregator');
require(libPath + "/constants");

// Used for generating random record bodies
var ALPHABET = 'abcdefghijklmnopqrstuvwxyz';

function randomRange(min, max) {
	return Math.trunc(Math.random() * (max - min)) + min;
}

function getRandomRecord(seqNum, desiredLen) {
	seqNum = seqNum !== undefined ? seqNum : 0;
	desiredLen = desiredLen !== undefined ? desiredLen : 50;

	var pk = (1.0 * Math.random()).toString().replace('.', '');
	var ehk = (1.0 * Math.random()).toString().replace('.', '');
	
	while (ehk[0] === '0' && ehk.length > 0) {
		ehk = ehk.substring(1);
	}

	var data = 'RECORD ' + seqNum.toString() + ' ';
	while (data.length < desiredLen - 1) {
		data += ALPHABET[randomRange(0, ALPHABET.length)]
	}
	data += '\n'

	return {
		'PartitionKey' : pk,
		'ExplicitHashKey' : ehk,
		'Data' : data
	};
}

function getKinesisClient(region_name) {
	return new AWS.Kinesis({
		'region' : region_name
	});
}

function sendRecord(kinesisClient, streamName, aggRecord) {
	params = {
		'StreamName' : streamName,
		'PartitionKey' : aggRecord.PartitionKey,
		'ExplicitHashKey' : aggRecord.ExplicitHashKey,
		'Data' : aggRecord.Data
	};

	var pk = aggRecord.PartitionKey;
	var ehk = aggRecord.ExplicitHashKey;

	console.log('Submitting record with PK=' + pk + ' EHK=' + ehk
			+ ' NumBytes=' + Buffer.byteLength(params.Data,'binary'));

	try {
		kinesisClient.putRecord(params, function(err, data) {
			if (err) {
				console.log('Transmission FAILED: ' + err);
				return;
			}

			console.log('Completed record with PK=' + pk + ' EHK=' + ehk);
		});
	} catch (err) {
		console.log('Transmission FAILED: ' + err);
	}
}

/*******************************************************************************
 * For setting AWS credentials, see:
 * https://docs.aws.amazon.com/AWSJavaScriptSDK/guide/node-configuring.html
 ******************************************************************************/

var RECORD_SIZE_BYTES = 1024;
var RECORDS_TO_TRANSMIT = 1024;

if (process.argv.length != 4) {
	console.log("USAGE: node kinesis_publisher.js <stream name> <region>");
	process.exit(1);
}

var streamName = process.argv[2];
var region = process.argv[3];

var kinesisClient = getKinesisClient(region);

// create semi-random records to transmit
console.log('Creating ' + RECORDS_TO_TRANSMIT + ' records...');
var recordsToSend = [];
for (i = 1; i <= RECORDS_TO_TRANSMIT; i++) {
	var rec = getRandomRecord(i, RECORD_SIZE_BYTES);
	recordsToSend.push(rec);
}

// aggregate all the records
console.log('Aggregating...');
var aggregator = new RecordAggregator();
aggregator.aggregateRecords(recordsToSend, false, null,
		function(err, aggRecord) {
			sendRecord(kinesisClient, streamName, aggRecord);
		});

// flush out any remaining records we haven't handled yet
console.log('Flushing...');
aggregator.flushBufferedRecords(function(err, aggRecord) {
	sendRecord(kinesisClient, streamName, aggRecord);
});
